packages/plugins/apollo-usage-report/src/index.ts

```diff
@@ -0,0 +1,172 @@
+              level,
+              (...args: unknown[]) => log('[ApolloUsageReport]', ...args),
+            ]),
+          ) as YogaLogger;
+
+          if (!apiKey) {
+            throw new Error(
+              `[ApolloUsageReport] Missing API key. Please provide one in plugin options or with 'APOLLO_KEY' environment variable.`,
+            );
+          }
+
+          if (!graphRef) {
+            throw new Error(
+              `[ApolloUsageReport] Missing Graph Ref. Please provide one in plugin options or with 'APOLLO_GRAPH_REF' environment variable.`,
+            );
+          }
+
+          logger.debug('using', { apiKey, graphRef });
+        },
+        async onSchemaChange({ schema }) {
+          if (schema) {
+            schemaId = await hashSHA256(printSchema(schema));
+          }
+        },
+        onParse() {
+          return ({ result, context: { request, params } }) => {
+            const ctx = ctxForReq.get(request);
+            if (!ctx) {
+              logger.debug('ctx not found during parsing');
+              return;
+            }
+
+            const { operationName } = params;
+            const signature = defaultUsageReportingSignature(result, operationName || '');
+            ctx.operationKey = `# ${operationName || '-'}\n${signature}`;
+            ctx.schemaId = schemaId;
+          };
+        },
+
+        async onResultProcess(args) {
+          const ctx = ctxForReq.get(args.request);
+          if (!ctx?.operationKey || !ctx?.schemaId) {
+            logger.debug('ctx not found during result processing:', ctx);
+            return;
+          }
+
+          // TODO: Handle async iterables ?
+          if (isAsyncIterable(args.result)) {
+            logger.debug('async iterable results not implemented for now');
+            return;
+          }
+
+          fetch(endpoint, {
+            method: 'POST',
+            headers: {
+              'content-type': 'application/protobuf',
+              // The presence of the api key is already checked at Yoga initialization time
+              // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+              'x-api-key': apiKey!,
+              accept: 'application/json',
+            },
+            body: Report.encode({
+              header: {
+                graphRef,
+                executableSchemaId: ctx.schemaId,
+              },
+              operationCount: 1,
+              tracesPerQuery: {
+                [ctx.operationKey]: {
+                  trace: [ctx.trace],
+                },
+              },
+            }).finish(),
+          }).then(async response => {
+            if (response.ok) {
+              logger.debug('Traces sent:', await response.text());
+            } else {
+              logger.error('Failed to send trace:', await response.text());
+            }
+          });
+        },
+      });
+    },
+  };
+}
+
+export async function hashSHA256(
```

Comment:
This is used some other places as well. Maybe we can export it from graphql-yoga package directly to re-use it in different places. What do you think?

Reply:
Yes, I was not sure where to put this, so I temporally copy pasted it